### PR TITLE
Backup & Restore Module

### PR DESCRIPTION
**Pull request by thybag**
Hello,

Hopefully I'm not being to annoying putting in all these random pull requests. 

Had a play with creating a 1 click backup & restore module for RES (as described in https://github.com/honestbleeps/Reddit-Enhancement-Suite/issues/300 ). Currently its a bit rough around the edges + I've yet to fully put it through its paces in terms of cross browser testing, but i thought i may as well put in a pull for visibility even if its decided eventually not to include the functionality (due to cross browser issues/support headaches it could cause).

It utilities- the data URI's for downloading the file (just stringified JSON) + the filereader API to read the file back in. Everything appears to be working in chrome.

### CODE REVIEW

**Comment by thybag** [comment]
If there's any interest/movement on this I'd be happy to merge it all up to date / check its all happy with the newer versions :+1:

**Comment by allthefoxes** [comment]
I am always going to be interested in having this.

File: lib/modules/backupAndRestore.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+			},
+			restore: {
+				type: 'button',
+				text: 'Restore',
+				callback: null,
+				description: 'Restore a backup of your RES settings. Warning: This will overwrite your current settings.'
+			}
+		},
+		description: 'Backup and restore your Reddit Enhancement Suite settings.',
+		isEnabled: function() {
```
Comment (andytuba) [inline lib/modules/backupAndRestore.js]: You can delete isEnabled, isMatchURL, and include

File: lib/modules/backupAndRestore.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+			return RESConsole.getModulePrefs(this.moduleID);
+		},
+		isMatchURL: function() {
+			return RESUtils.isMatchURL(this.moduleID);
+		},
+		include: [
+			'all'
+		],
+		go: function() {
```
Comment (andytuba) [inline lib/modules/backupAndRestore.js]: this should be loadDynamicOptions now that it exists

**Comment by andytuba** [comment]
Could you test this on Firefox too, @thybag? This would be a good olive branch for all the Firefox users who report data loss.

Nice to have features for the future:

* notification to run a backup, cooldown of 2 weeks
* ^ smart cooldown of "X [option, default = 14] days since your last backup" easing into every day.  (option, default enabled)
* select certain modules / data to export -- perhaps as named sets, or using a customFilters-like interface
* ^ and run migration on imported data if needed, which requires tagging each piece of data with the RESOptionsVersion number

**Comment by thybag** [comment]
Just synced it all up with master again (plus made the suggested changes).
Seems to work fine (managed to import a settings dump i made  with the first version of this, then rollback to an export i made with this version) :)

**Comment by thybag** [comment]
Think all issues in travis that relate to `backupAndRestore` should now be resolved :+1:

**Pull request closed without merging**